### DISCUSSION TITLE
Why are unit conversion results are different when using the Quantity() constructor?

### DISCUSSION THREAD

**Discussion by cdwilson** [hgrecco/pint#2105] [Q&A]
I was surprised to find that converting `1 inch` to `thou` using Decimal numbers produced a result with incorrect precision.

```python
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry(non_int_type=Decimal)
>>> d1 = Decimal("1.0") * ureg.inch
>>> d1.to(ureg.thou).magnitude
Decimal('999.9999999999999999999999996')
```

However, if I use the Quantity() constructor, I get the correct precision I would expect using a Decimal number.

```python
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry(non_int_type=Decimal)
>>> d2 = Quantity(Decimal("1.0"), ureg.inch)
>>> d2.to(ureg.thou).magnitude
Decimal('1000.00')
```

Can someone explain why these produce different results? What is the difference between creating the quantity with `Decimal("1") * ureg.inch` vs. `Quantity(Decimal(1), ureg.inch)`?

**Comment by cdwilson** [comment]
It looks like there were a couple issues with the examples I originally posted above.

First, when I tried the 2nd example again today in a fresh repl, I realized it doesn't work as-is. After fixing it, I get the following error:

```python
>>> from decimal import Decimal
>>> from pint import Quantity, UnitRegistry
>>> ureg = UnitRegistry(non_int_type=Decimal)
>>> d2 = Quantity(Decimal("1.0"), ureg.inch)
>>> d2.to(ureg.thou).magnitude
Traceback (most recent call last):
[...]
TypeError: unsupported operand type(s) for ** or pow(): 'float' and 'decimal.Decimal'
>>> 
```

I think when I was originally trying this out I had accidentally typed `ureg = UnitRegistry()` in the repl before running the `d2` example:

```python
>>> from decimal import Decimal
>>> from pint import UnitRegistry, Quantity
>>> ureg = UnitRegistry()
>>> d2 = Quantity(Decimal("1.0"), ureg.inch)
>>> d2.to(ureg.thou).magnitude
Decimal('1000.00')
```

The second issue is that I was importing `Quantity` directly from the `pint` module instead of from the `UnitRegistry` instance. If I change the `d2` example to use `ureg.Quantity` then I get the same results as the `d1` example:

```python
>>> from decimal import Decimal
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry(non_int_type=Decimal)
>>> d2 = ureg.Quantity(Decimal("1.0"), ureg.inch)
>>> d2.to(ureg.thou).magnitude
Decimal('999.9999999999999999999999996')
```

Couple questions:
1) Why does `ureg.Quantity` behave differently from `from pint import Quantity`?
2) I'm still confused as to why there is a loss of precision when using Decimal numbers in the last example. Is there a way to rewrite this so that the conversion does the equivalent of `Decimal("1.0") * 1000`, i.e. the conversion returns `Decimal('1000.0')`?

**Reply by andrewgsavage** [reply]
1. the ureg you created uses `non_int_type=Decimal`, whereas `from pint import Quantity` uses the application registry that has that as float
2. not sure. I'd guess there's a float used somewhere between parsing the unit file and calcuating conversion factor. that'd need looking into

**Reply by cdwilson** [reply]
@andrewgsavage thank you for your reply!

This is my first time using pint so I wasn't sure if I was doing something incorrectly or if this was a bug in the way Decimals were being handled in the conversion. From your reply, it sounds like this is not expected behavior and there shouldn't be a conversion to a float in process. Let me know if that's correct and I will file an issue for this.

**Reply by andrewgsavage** [reply]
yea you can create an issue

you may be able to use floats and formatting so you dont need `Decimal`, eg
```
>>> from pint import UnitRegistry
>>> ureg = UnitRegistry()
>>> d2 = ureg.Quantity(1-1e-12, ureg.inch)
>>> d3=d2.to(ureg.thou)
>>> f"{d3:.2f~#P}"
'1000.00 th'
```
https://pint.readthedocs.io/en/stable/user/formatting.html
